community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/command/NeoStoreTransactionApplier.java

```diff
@@ -97,34 +158,39 @@ else if ( command.getRelId() != -1 )
         // track the dynamic value record high ids
         // update store
         neoStores.getPropertyStore().updateRecord( command.getAfter() );
+        freeIds( neoStores.getPropertyStore(), command.getAfter() );
```

Comment:
In 3.0, this bit of code will be below the `StorageEngine` in the stack of abstractions. Perhaps that's a good thing since only the StorageEngine knows how many records any entity takes up. Perhaps the StorageEngine needs to grow some methods so it can be informed about transactions starting and stopping. That way, all of this id-freeing machinery can be hidden inside of it.